File: geocoder/src/main/java/com/prolificinteractive/geocoder/GeocoderBuilder.java
package com.prolificinteractive.geocoder;

import java.util.ArrayList;
import java.util.List;

public class GeocoderBuilder {
  private final List<GeocodingApi> geocodingApis = new ArrayList<>();
  private Downloader.Factory downloaderFactory;
  private SwitchPolicy switchPolicy;

  public GeocoderBuilder addGeocodingApi(final GeocodingApi geocodingApi) {
    if (geocodingApi != null) {
      this.geocodingApis.add(geocodingApi);
    }

    return this;
  }

  public Geocoder build() {
    return new Geocoder(downloaderFactory, switchPolicy, geocodingApis);
  }

  public GeocoderBuilder setDownloaderFactory(final Downloader.Factory downloaderFactory) {
    this.downloaderFactory = downloaderFactory;
    return this;
  }

  public GeocoderBuilder setSwitchPolicy(final SwitchPolicy switchPolicy) {
    this.switchPolicy = switchPolicy;
    return this;
  }
}

File: geocoder/src/main/java/com/prolificinteractive/geocoder/GeocoderBuilder.kt
package com.prolificinteractive.geocoder

import com.prolificinteractive.geocoder.model.Address
import java.util.ArrayList

class GeocoderBuilder {
  private val geocodingApis = ArrayList<GeocodingApi>()
  private var downloaderFactory: Downloader.Factory? = null
  private var switchPolicy: SwitchPolicy = object : SwitchPolicy {
    override fun shouldSwitch(geocoderName: String, addresses: List<Address>): Boolean {
      return false
    }
  }

  fun addGeocodingApi(geocodingApi: GeocodingApi?): GeocoderBuilder {
    if (geocodingApi != null) {
      this.geocodingApis.add(geocodingApi)
    }

    return this
  }

  fun build(): Geocoder {
    return Geocoder(downloaderFactory, switchPolicy, geocodingApis)
  }

  fun setDownloaderFactory(downloaderFactory: Downloader.Factory): GeocoderBuilder {
    this.downloaderFactory = downloaderFactory
    return this
  }

  fun setSwitchPolicy(switchPolicy: SwitchPolicy): GeocoderBuilder {
    this.switchPolicy = switchPolicy
    return this
  }
}
